Build process android

Sumenia opened the pull request:
This is the build process working for android Marshmallow and Nougat.
findLug doesn’t work within gradle so having lugdunum library install
in android/app/src/main/lugdunum is mandatory

Due to dynamic library dependancy policy in Android M, building
Lugdunum in static is mandatory.

also :
I removed weird broken ascii spaces across the project
Fixed the README.md

Commit:
build process android

This is the build process working for android Marshmallow and Nougat.
findLug doesn’t work within gradle so having lugdunum library install
in android/app/src/main/lugdunum is mandatory

Due to dynamic library dependancy policy in Android M, building
Lugdunum in static is mandatory.

also :
I removed weird broken ascii spaces across the project
Fixed the README.md

saveman71 commented on include/lug/System/Exception.hpp:
```diff
@@ -9,9 +9,9 @@
 namespace System {
 
-class LUG_SYSTEM_API Exception: public std::exception {
+class LUG_SYSTEM_API Exception: public std::exception  {
```
Two spaces?

saveman71 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
+    # build native_app_glue as a static lib
+    set(APP_GLUE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/android/app/src/main/cpp/native_app_glue)
+    include_directories(${APP_GLUE_DIR})
+    add_library( app-glue STATIC ${APP_GLUE_DIR}/android_native_app_glue.c)
```
Space after `(`

saveman71 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
+    list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/../../cmake/modules")
 endif()
 
-# project name
```
Why did you remove those lines?

saveman71 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
+                          ${LUG_DIR}/lib/liblug-system-s-d.a)
+    else()
+    target_link_libraries(sample
+                              app-glue
```
Indentation issue

saveman71 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
-# determine the build type
-if(NOT CMAKE_BUILD_TYPE)
-    set(CMAKE_BUILD_TYPE Debug CACHE STRING "Choose the type of build (Debug or Release)" FORCE)
+if(NOT ANDROID)
```
Can you merge the two `NOT ANDROID` blocks in a single one?

saveman71 commented on samples/base/main.cpp:
```diff
@@ -1,5 +1,40 @@
-#include <lug/Config.hpp>
+#if defined(__ANDROID__)
+ #include <android/log.h>
```
4 spaces indent

saveman71 commented on samples/base/main.cpp:
```diff
@@ -1,5 +1,40 @@
+ #include <android/log.h>
+ #define LOGV(...)  ((void)__android_log_print(ANDROID_LOG_VERBOSE, "lugdunum sample", __VA_ARGS__))
+ #include "android_native_app_glue.h"
+ #include "Lug/System/Exception.hpp"
```
Use `<>` for lug headers

saveman71 commented on samples/base/main.cpp:
```diff
@@ -1,5 +1,40 @@
+#if defined(__ANDROID__)
+    app_dummy();
+    while (1)
+    {
```
`{` goes at the end of previous lines. Applies to the `while` and the `if`s below too.

saveman71 commented on src/lug/System/Exception.cpp:
```diff
@@ -1,8 +1,11 @@
 
-lug::System::Exception::Exception(const char *typeName, const std::string &description, const char* file, const char* function, uint32_t line)
-    : _typeName{typeName}, _description{description}, _file{file}, _function{function}, _line{line} {}
+lug::System::Exception::Exception(const char *typeName, const std::string &description, const char* file, const char* function, uint32_t line) noexcept
```
Why? (4 spaces ident is lost, and the guidelines are not respected anymore)

saveman71 commented on samples/base/main.cpp:
```diff
@@ -1,5 +1,40 @@
-#include <lug/Config.hpp>
+#if defined(__ANDROID__)
+ #include <android/log.h>
+ #define LOGV(...)  ((void)__android_log_print(ANDROID_LOG_VERBOSE, "lugdunum sample", __VA_ARGS__))
```
You define `LOGV` but do not use it

saveman71 commented on samples/base/main.cpp:
```diff
@@ -1,5 +1,40 @@
+    int main()
+#endif
+{
+#if defined(__ANDROID__)
```
You should define `LUG_SYSTEM_ANDROID` in `include/lug/Config.hpp` instead of using directly `__ANDROID__`

Commit:
adding gitignore rules

adding unwanted file in the gitignore

Commit:
android 24 fixed

fixing indentation problem and android-24 target not working properly

Quentin01 commented on CMakeLists.txt:
```diff
@@ -13,9 +13,31 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+# determine the build type
 lug_set_option(CMAKE_BUILD_TYPE Release STRING "Choose the type of build (Debug or Release)")
 
+if(ANDROID)
```
Why not using the same check than in `Config.cmake`?

Quentin01 replied:
Maybe some vertical spaces in this block?

Quentin01 commented on CMakeLists.txt:
```diff
@@ -13,9 +13,31 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
 lug_set_option(CMAKE_BUILD_TYPE Release STRING "Choose the type of build (Debug or Release)")
 
+if(ANDROID)
+    set(ANDROID_ABI arm64-v8a)
```
So, the user haven't the choice?

Sumenia replied:
Nope. Android N and Vulkan are both arm64-v8a. There is no new arm architecture at this moment, not even planned. i guess we're good

Quentin01 commented on CMakeLists.txt:
```diff
@@ -13,9 +13,31 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+    set(ANDROID_TOOLCHAIN clang)
+    message(STATUS "ANDROID_TOOLCHAIN : ${ANDROID_TOOLCHAIN} ")
+    lug_set_option(ANDROID_PLATFORM android-24 STRING "Target specific android version target (default : android-24)")
+    if(NOT ANDROID_PLATFORM MATCHES android-2[3-4])
```
In the future with `android-25`, what will you do?

Sumenia replied:
We have a full year ahead of us to even starting thinking to handle that.
I can modify for [3-9] if you want to have 5 more years planned

Quentin01 commented on CMakeLists.txt:
```diff
@@ -13,9 +13,31 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+        message(FATAL_ERROR "Invalid android version target : ${ANDROID_PLATFORM}")
+    endif()
+    message(STATUS "ANDROID_PLATFORM : ${ANDROID_PLATFORM}")
+    if(ANDROID_PLATFORM MATCHES android-23)
```
Don't you want to also use `c++_static` in case of platform 23?

Quentin01 commented on cmake/modules/FindLUG.cmake:
```diff
@@ -165,6 +166,7 @@ endforeach()
 
 if (LUG_FOUND)
     message(STATUS "Found Lugdunum in ${LUG_INCLUDE_DIR}")
+    message(STATUS "Found Lugdunum in ${LUG_LIBRARIES}")
```
Maybe don't use the message twice?

Sumenia replied:
Which one do we change ? The first one by "Lugdunum headers" ?

Quentin01 replied:
Yes, "Lugdunum headers" seems good.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
-cmake_minimum_required(VERSION 3.1)
-
-# determine the build type
-if(NOT CMAKE_BUILD_TYPE)
```
For what reason do you have remove that? You may need this for the Android Build

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
 project(base)
 
-# define the executable
-set(EXECUTABLE_NAME "base")
```
Why don't you use this one for the Android build too?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+    # build native_app_glue as a static lib
+    set(APP_GLUE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/android/app/src/main/cpp/native_app_glue)
+    include_directories(${APP_GLUE_DIR})
+    add_library( app-glue STATIC ${APP_GLUE_DIR}/android_native_app_glue.c)
```
No space after `(`

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+    add_library( app-glue STATIC ${APP_GLUE_DIR}/android_native_app_glue.c)
+
+    # temporary workaround for findLug not working with gradle
+    set(LUG_DIR ${CMAKE_CURRENT_SOURCE_DIR}/android/app/src/main/lugdunum)
```
Maybe use the same convention, so `LUG_ROOT` as everywhere

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+    set(LUG_DIR ${CMAKE_CURRENT_SOURCE_DIR}/android/app/src/main/lugdunum)
+    include_directories(${LUG_DIR}/include)
+
+    add_library( # Sets the name of the library : must match the name defined in AndroidManifest.xml
```
No space after `(`. I doubt of the usefulness of those comments except the one for `AndroidManifest.xml`.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+                              app-glue
+                              log
+                              android
+                              ${LUG_DIR}/lib/liblug-system-s-d.a)
```
So, you only use the debug library.
Use CMAKE_BUILD_TYPE to choose which one you want.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+                              ${LUG_DIR}/lib/liblug-system-s-d.a)
+
+    elseif(ANDROID_NATIVE_API_LEVEL EQUAL 24)
+        add_library(lug-system-d
```
So, you only use the debug library.
Use CMAKE_BUILD_TYPE to choose which one you want.

Quentin01 replied:
And why all those newlines? That seems clear without.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+                              PROPERTIES IMPORTED_LOCATION
+
+                              # Provides the path to the library you want to import.
+                              ${LUG_DIR}/lib/arm64-v8a/liblug-system-d.so)
```
So, you only use the debug library.
Use CMAKE_BUILD_TYPE to choose which one you want.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+                    SHARED
+                    IMPORTED)
+
+        set_target_properties(# Specifies the target library.
```
Same question, why those newlines?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -1,22 +1,73 @@
+                              app-glue
+                              log
+                              android
+                              lug-system-d)
```
So, you only use the debug library.
Use CMAKE_BUILD_TYPE to choose which one you want.

Quentin01 reviewed (changes requested):
Where is lug-main?

Commit:
PR review fixed

FindLUG now work with Android.
Deleted add_library and set_target_properties in the Cmake of the
Sample because it was not needed
Automatically package all .so in the APK when needed

saveman71 committed:
Add CMakeLists for Main

Commit:
AndroidMain implementation

i deleted the android native app glu library from NDK in favour of our
own implementation.

saveman71 committed:
Merge branch 'dev' into featue-build-process-android

Commit:
better name for the android dynamic lib that match the executable name

saveman71 committed:
Fix CMakelists.txt

Commit:
vertical space + move MainAndroid in /Main

saveman71 committed:
Add lug-main

saveman71 committed:
Revert Core

Commit:
fix android build

Commit:
findLug now give the MainAndroid.a

Commit:
lug-main.a is now working correctly (bug fix from undefined symbol )

Commit:
fix CXX_FLAG for android

small fix, was CMAKE_CXX_FLAGS and not ${CMAKE_CXX_FLAGS}

Fix formatting

saveman71 committed:
Convert everything to OOP

Quentin01 commented on CMakeLists.txt:
```diff
@@ -13,9 +13,43 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+    message(STATUS "ANDROID_ABI : ${ANDROID_ABI}")
+
+    set(ANDROID_TOOLCHAIN clang)
+    message(STATUS "ANDROID_TOOLCHAIN : ${ANDROID_TOOLCHAIN} ")
```
Space at the end of the message

Quentin01 commented on cmake/modules/FindLUG.cmake:
```diff
@@ -50,6 +50,7 @@ endif()
 set(FIND_LUG_PATHS
     ${LUG_ROOT}
     $ENV{LUG_ROOT}
+    ${ANDROID_NDK}/sources/lugdunum
```
Maybe also use the environnement variable?

Sumenia replied:
didn't we did this together ? That was the problem why FingLug failed when invoked by Android Studio

Quentin01 commented on include/lug/Main/Export.hpp:
```diff
@@ -0,0 +1,13 @@
+#pragma once
```
Useless file

Quentin01 commented on include/lug/Main/MainAndroid.hpp:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <android/native_activity.h>
```
Bad order of inclusion, include thread and mutex before.

Quentin01 commented on include/lug/Main/MainAndroid.hpp:
```diff
@@ -0,0 +1,49 @@
+    void* _savedState;
+    size_t _savedStateSize;
+
+    std::mutex _mutex;
```
Explicit name for the mutex? What is it purpose?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,44 @@ set(CXX_STANDARD_REQUIRED ON)
+   endif()
 elseif(LUG_COMPILER_GCC OR LUG_COMPILER_CLANG)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+   list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
```
And "-Werror" ?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,44 @@ set(CXX_STANDARD_REQUIRED ON)
-find_package(LUG REQUIRED system)
-if(LUG_FOUND)
-    include_directories(${LUG_INCLUDE_DIR})
+find_package(LUG REQUIRED)
```
Link with system too!

saveman71 replied:
And why? Currently the sample doesn't use any functionality from system.

Quentin01 replied:
This sample is just a test and a demonstration for the user, so we can make it obvious that he needs to tell which librairies he wants.

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,44 @@ set(CXX_STANDARD_REQUIRED ON)
-if(LUG_FOUND)
-    include_directories(${LUG_INCLUDE_DIR})
+find_package(LUG REQUIRED)
+if(NOT LUG_FOUND)
```
Not usefull, this error will already happens in the FindLUG.cmake

Quentin01 commented on src/lug/CMakeLists.txt:
```diff
@@ -24,11 +24,20 @@ if(LUG_COMPILER_MSVC)
     endif()
 elseif(LUG_COMPILER_GCC OR LUG_COMPILER_CLANG)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
```
And "-Werror"?

Quentin01 commented on src/lug/CMakeLists.txt:
```diff
@@ -24,11 +24,20 @@ if(LUG_COMPILER_MSVC)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
+    if(LUG_OS_ANDROID)
+        list(APPEND CMAKE_CXX_FLAGS "-std=c++14")
```
Maybe just use that method for the all project?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,44 @@ set(CXX_STANDARD_REQUIRED ON)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+   list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
+    if(LUG_OS_ANDROID)
+        list(APPEND CMAKE_CXX_FLAGS "-std=c++14")
```
Maybe just use that method for the all project?

Quentin01 commented on src/lug/Main/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+
+# Include Main cpp files depending on user os
+if(LUG_OS_WINDOWS)
+    list(APPEND SRC ${SRCROOT}/MainWindows.cpp)
```
No MainWindows for the moment, so ...?
Make it or just remove the support for Windows now (because that will break the build on Windows)

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+        _savedStateSize = 0;
+    }
+    _mutex.unlock();
+    delete this;
```
I don't think deleting itself is a great idea

saveman71 replied:
How would you implement it then? We have to alloc it to java in ANativeActivity_onCreate and it is guaranteed that shutdowApplication is in fact, putting the application down for real. If we don't delete ourselves, we leak some memory

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+
+namespace lug {
+namespace Main {
+
```
Double vertical space?

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+    activity->callbacks->onDestroy = &LugAndroidApp::onDestroy;
+
+    if (savedState != nullptr) {
+        _savedState = ::operator new(savedStateSize);
```
Maybe allocate an array of uint8_t?

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+void LugAndroidApp::shutdowApplication() {
+    _mutex.lock();
+    if (_savedState != nullptr) {
+        ::operator delete(_savedState);
```
If you change the allocation, use `delete[]`

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+
+
+LugAndroidApp::LugAndroidApp(ANativeActivity* activity, void* savedState, size_t savedStateSize)
+    : _activity(activity), _savedState(nullptr), _savedStateSize(0) {
```
Why not using default constructor for `_savedState` and `_savedStateSize`?

Quentin01 replied:
Brackets instead of parentheses

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+    if (savedState != nullptr) {
+        _savedState = ::operator new(savedStateSize);
+        _savedStateSize = savedStateSize;
+        memcpy(_savedState, savedState, savedStateSize);
```
Prefix `memcpy` by `std::`

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    std::thread t(&LugAndroidApp::startApplication, this);
+    t.detach();
```
Isn't there a better solution?

saveman71 replied:
After this function the thread is completely independent (Should _not_ be joined), and we hand the main thread back to java. Do _you_ have a better solution in mind?

Quentin01 commented on src/lug/Main/MainAndroid.cpp:
```diff
@@ -0,0 +1,126 @@
+void LugAndroidApp::onLowMemory(ANativeActivity* activity) {
+
+}
+
```
Double vertical space?

Quentin01 commented on include/lug/Main/MainAndroid.hpp:
```diff
@@ -0,0 +1,49 @@
+#include <android/configuration.h>
+#include <thread>
+#include <mutex>
+#include <lug/Main/Export.hpp>
```
Useless include

Quentin01 commented on include/lug/Main/MainAndroid.hpp:
```diff
@@ -0,0 +1,49 @@
+
+#include <android/native_activity.h>
+#include <android/configuration.h>
+#include <thread>
```
Don't include thread if you aren't using it in the header! (Include it in the source file)

Commit:
global gradle solution

i moved the global gradle in /sample/compiler/android

each sample is going to be a project of that global gradle

Commit:
Misc fixes for review

prefix std:: for memcpy

removing mainWindow

This branch is about Android Build System we should not handle
WindowMain just yet

removing noexcept

This branch is about Android build system. When this one is merged, a
new branch about testing feature on Android will be created

Sumenia commented on include/lug/System/Exception.hpp:
```diff
@@ -12,7 +12,7 @@ namespace System {
 class LUG_SYSTEM_API Exception: public std::exception {
 public:
-    Exception(const char *typeName, const std::string &description, const char* file, const char* function, uint32_t line);
+    Exception(const char *typeName, const std::string &description, const char* file, const char* function, uint32_t line) noexcept;
```
@Quentin01 should i fix this in this branch or create another branch about this issue and further tests of Exception on Android when this one is merged ?

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,41 @@ set(CXX_STANDARD_REQUIRED ON)
+   endif()
 elseif(LUG_COMPILER_GCC OR LUG_COMPILER_CLANG)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+   list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
```
Missing one space

Quentin01 commented on samples/base/CMakeLists.txt:
```diff
@@ -20,18 +19,41 @@ set(CXX_STANDARD_REQUIRED ON)
-    list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+   list(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra -Werror")
+    if(LUG_OS_ANDROID)
+        list(APPEND CMAKE_CXX_FLAGS "-std=c++14")
```
Use that for all the project and remove `CMAKE_CXX_STANDARD 14`

Quentin01 commented on include/lug/Main/MainAndroid.hpp:
```diff
@@ -0,0 +1,42 @@
+#pragma once
```
Does the user needs that file? If not move it to the src directory.
